exchanges/exchanges.go

```diff
@@ -71,15 +160,195 @@ func Tokens() []string {
+type DepthPoint struct {
+	Quantity float64 `json:"quantity"`
+	Price    float64 `json:"price"`
+}
+
+// DepthData is an exchanges order book for use in a depth chart.
+type DepthData struct {
+	Time time.Time    `json:"time"`
+	Bids []DepthPoint `json:"bids"`
+	Asks []DepthPoint `json:"asks"`
+}
+
+// IsFresh will be true if the data is older than depthDataExpiration.
+func (depth *DepthData) IsFresh() bool {
+	if time.Now().Sub(depth.Time) < depthDataExpiration {
+		return true
+	}
+	return false
+}
+
+// Candlestick is the record of price change over some bin width of time.
+type Candlestick struct {
+	High   float64   `json:"high"`
+	Low    float64   `json:"low"`
+	Open   float64   `json:"open"`
+	Close  float64   `json:"close"`
+	Volume float64   `json:"volume"`
+	Start  time.Time `json:"start"`
+}
+
+// Candlesticks is a slice of CandleStick.
+type Candlesticks []Candlestick
+
+// returns the start time of the last Candlestick, else the zero time,
+func (sticks Candlesticks) time() time.Time {
+	if len(sticks) > 0 {
+		return sticks[len(sticks)-1].Start
+	}
+	return time.Time{}
+}
+
+// Checks whether the candlestick data for the given bin size is up-to-date.
+func (sticks Candlesticks) needsUpdate(bin candlestickKey) bool {
+	if len(sticks) == 0 {
+		return true
+	}
+	lastStick := sticks[len(sticks)-1]
+	if time.Now().After(lastStick.Start.Add(bin.duration() * 2)) {
+		return true
+	}
+	return false
+}
+
 // ExchangeState is the simple template for a price. The only member that is
 // guaranteed is a price. For Decred exchanges, the volumes will also be
 // populated.
 type ExchangeState struct {
-	Price      float64 `json:"price"`
-	BaseVolume float64 `json:"base_volume,omitempty"`
-	Volume     float64 `json:"volume,omitempty"`
-	Change     float64 `json:"change,omitempty"`
-	Stamp      int64   `json:"timestamp,omitempty"`
+	Price        float64                         `json:"price"`
+	BaseVolume   float64                         `json:"base_volume,omitempty"`
+	Volume       float64                         `json:"volume,omitempty"`
+	Change       float64                         `json:"change,omitempty"`
+	Stamp        int64                           `json:"timestamp,omitempty"`
+	Depth        *DepthData                      `json:"depth,omitempty"`
+	Candlesticks map[candlestickKey]Candlesticks `json:"candlesticks,omitempty"`
+}
+
+func (state *ExchangeState) copy() *ExchangeState {
+	newState := &ExchangeState{
+		Price:      state.Price,
+		BaseVolume: state.BaseVolume,
+		Volume:     state.Volume,
+		Change:     state.Change,
+		Stamp:      state.Stamp,
+		Depth:      state.Depth,
+	}
+	if state.Candlesticks != nil {
+		newState.Candlesticks = make(map[candlestickKey]Candlesticks)
+		for bin, sticks := range state.Candlesticks {
+			newState.Candlesticks[bin] = sticks
+		}
+	}
+	return newState
+}
+
+// Create a new exchange state that projects the fields from `top` over a copy
+// of the receiver.
```

Comment:
For this function it's probably valuable to document the intend more thoroughly.

Reply:
I think I'm going to rewrite this `project` function. It got a little convoluted.

Reply:
Thanks. I had that all messed up.